src/array/cuda/gather_mm.cu

```diff
@@ -0,0 +1,598 @@
+        const DType *A_data = A.Ptr<DType>();
+        const DType *B_data = B.Ptr<DType>();
+        int out_len = B->shape[1];  // cols of B
+        int in_len = A->shape[1];  // cols of A
+        if (!thr_entry->cublas_handle)
+            CUBLAS_CALL(cublasCreate(&(thr_entry->cublas_handle)));
+        CUBLAS_CALL(cublasSetStream(thr_entry->cublas_handle,
+            thr_entry->stream));
+        IdType tot_num_rows = A->shape[0];
+        const int ntx = 128;
+        const int warp_size = 32;
+        const int nbx =  ((tot_num_rows * warp_size + ntx - 1) / ntx);
+        const dim3 nblks(nbx);
+        const dim3 nthrs(ntx);
+        CUDA_KERNEL_CALL((gatherMMKernel<IdType, DType>),
+            nblks, nthrs, 0, thr_entry->stream,
+            static_cast<DType*>(A->data),
+            static_cast<DType*>(B->data),
+            static_cast<DType*>(C->data),
+            static_cast<IdType*>(idx_a->data),
+            static_cast<IdType*>(idx_b->data),
+            tot_num_rows,
+            in_len, out_len);
+    });
+}
+
+/* \brief Implementation of GatherMM operator. The indices of A (or B or C)
+ * are looked up from idx_a (or idx_b or idx_c) when defined.
+ */
+template <int XPU, typename IdType, int bits>
+void gatherMM_scatter(const NDArray A,
+              const NDArray B,
+              NDArray C,
+              const NDArray idx_a,
+              const NDArray idx_b,
+              const NDArray idx_c,
+              int num_rel, bool a_trans, bool b_trans) {
+    SWITCH_BITS(bits, DType, {
+        auto device = runtime::DeviceAPI::Get(A->ctx);
+        auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+        const IdType *idx_c_data = idx_c.Ptr<IdType>();
+        int out_len = B->shape[1];  // cols of B
+        int in_len = A->shape[1];  // cols of A
+        if (!thr_entry->cublas_handle)
+            CUBLAS_CALL(cublasCreate(&(thr_entry->cublas_handle)));
+        CUBLAS_CALL(cublasSetStream(thr_entry->cublas_handle,
+            thr_entry->stream));
+        DType* B_trans_data = nullptr;
+        if (b_trans) {
+            int64_t B_offset = 0;
+            const DType *B_data = B.Ptr<DType>();
+            in_len = B->shape[0]/num_rel;
+            B_trans_data = static_cast<DType*>(device->AllocWorkspace \
+                (B->ctx, B->shape[0] * B->shape[1] * sizeof(DType)));
+            // tranpose B per relation
+            for (int rel = 0; rel < num_rel; ++rel) {
+                _Transpose(thr_entry->cublas_handle, B_data + B_offset,
+                    B_trans_data + B_offset, in_len, out_len);
+                B_offset += in_len * out_len;
+            }
+            std::swap(in_len, out_len);
+        }
+        IdType tot_num_rows = A->shape[0];
+        const int ntx = 128;
+        const int warp_size = 32;
+        const int nbx =  ((tot_num_rows * warp_size + ntx - 1) / ntx);
+        const dim3 nblks(nbx);
+        const dim3 nthrs(ntx);
+        if (a_trans && idx_c_data) {
```

Comment:
I think you could ignore `a_trans` for now? since each row of A must be a vector.